File: 3594.minimum-time-to-transport-all-individuals.py
#
# @lc app=leetcode id=3594 lang=python3
#
# [3594] Minimum Time to Transport All Individuals
#

# @lc code=start
class Solution:
    def minTime(self, n: int, k: int, m: int, time: List[int], mul: List[float]) -> float:
        from itertools import combinations
        from math import floor
        
        # Calculate crossing time for a group based on current stage
        def crossing_time(group, stage):
            max_time = max(time[i] for i in group)
            return max_time * mul[stage]
        
        min_total_time = float('inf')  # Initialize minimum time as infinity
        
        # DFS function to explore possible solutions
        def dfs(at_base_camp, at_destination, current_stage, total_time):
            nonlocal min_total_time  # Access outer scope variable for min time tracking
            
            # If all individuals are at destination update minimum time if needed
            if len(at_destination) == n:
                min_total_time = min(min_total_time, total_time)
                return
            
            # Prune paths that exceed known minimum time already found
            if total_time >= min_total_time:
                return
            
            # Explore combinations of people moving from base camp within capacity limits (k)
            for group in combinations(at_base_camp, min(k, len(at_base_camp))):
                d = crossing_time(group, current_stage)  # Calculate forward trip time for this group
                new_stage = (current_stage + floor(d)) % m  # Compute new stage index post-trip
                new_at_destination = at_destination + list(group)  # Update destination list post-trip
                new_at_base_camp = [p for p in at_base_camp if p not in group]  # Update base camp list post-trip
                
                # Recursive DFS call with updated state after forward trip completion
                dfs(new_at_base_camp, new_at_destination, new_stage, total_time + d)
                
                # Handle return logic only if there are people left at base camp needing pickup later or more trips necessary to complete task.
                if not new_at_base_camp:
                    continue  # Skip rest of return logic since no one left behind for trips back home.
                    
                # Try returning each person individually back to base camp from destination for further trips if necessary.
                for p in new_at_destination:
                    r_d = crossing_time([p], new_stage)  # Calculate return trip time for this individual alone.
                    r_new_stage = (new_stage + floor(r_d)) % m  # Compute new stage index post-return trip.
                    r_new_at_destination = [x for x in new_at_destination if x != p]   # Update destination list excluding returning individual.
                    r_new_at_base_camp = new_at_base_camp + [p]    # Update base camp list including returning individual.
                    dfs(r_new_at_base_camp,r_new_at_destination,r_new_stage,total_time + d + r_d)   	          	         	         dfs(list(range(n)), [], 0, 0.0)  # Start DFS with all at base camp and zero initial time elapsed.	total_minimum_possible=float('inf')    return -1 if min_total_possible == float('inf') else min_total_possible# @lc code=end

File: 3594.minimum-time-to-transport-all-individuals.cpp
# @lc app=leetcode id=3594 lang=cpp
# [3594] Minimum Time to Transport All Individuals
#
# @lc code=start
#include <vector>
#include <algorithm>
#include <limits>
#include <unordered_map>
using namespace std;

class Solution {
public:
    double minTime(int n, int k, int m, vector<int>& time, vector<double>& mul) {
        unordered_map<int, double> memo; // Memoization map for storing states
        
        function<double(int,int,int)> dfs = [&](int mask, int stage, int remaining) -> double {
            if (remaining == 0) return 0; // All crossed over
            if (memo.count(mask)) return memo[mask];
            
            double min_time = numeric_limits<double>::max();
            
            // Try all combinations of up to k individuals using a bitmask approach
            for (int i = 0; i < (1 << n); ++i) { // Loop over all subsets of people
                if (__builtin_popcount(i) > k || (mask & i) != i) continue; // Skip invalid sets
                
                double max_cross_time = 0;
                for (int j = 0; j < n; ++j) {
                    if (i & (1 << j)) {
                        max_cross_time = max(max_cross_time, time[j] * mul[stage]);
                    }
                }
                int new_stage = (stage + static_cast<int>(floor(max_cross_time))) % m;
                double sub_result = dfs(mask ^ i, new_stage, remaining - __builtin_popcount(i));
                if (sub_result >= 0) {
                    min_time = min(min_time, max_cross_time + sub_result);
                }
            }
            
            return memo[mask] = (min_time == numeric_limits<double>::max()) ? -1 : min_time;
        };
        
        return dfs((1 << n) - 1, 0, n);
    }
};
# @lc code=end